generator/Data/XCB/Python/Parse.hs

```diff
@@ -326,12 +364,49 @@ structElemToPyPack :: String
                    -> TypeInfoMap
                    -> (String -> String)
                    -> GenStructElem Type
-                   -> Either (Maybe String, String) [(String, Expr ())]
+                   -> Either (Maybe String, String) [(String, [(Expr (), Maybe (Expr ()))])]
 structElemToPyPack _ _ _ (Pad i) = Left (Nothing, mkPad i)
 -- TODO: implement doc, switch, and fd?
 structElemToPyPack _ _ _ (Doc _ _ _) = Left (Nothing, "")
-structElemToPyPack _ _ _ (Switch _ _ _) = Left (Nothing, "")
 structElemToPyPack _ _ _ (Fd _) = Left (Nothing, "")
+structElemToPyPack ext m accessor (Switch n expr bitcases) =
+  let name = accessor n
+      cmp = xExpressionToPyExpr id expr
+      switch = map (mkSwitch cmp) bitcases
+  in Right $ [(name
+             , map (pkSwitch name ext m accessor) switch
+             )]
+    where
+      mkSwitch :: Expr ()
+               -> BitCase
+               -> (GenStructElem Type, Expr ())
+      mkSwitch cmp (BitCase _ bc_cmp elems) =
+        let cmp_val = xExpressionToPyExpr id bc_cmp
+            equality = BinaryOp (P.Equality ()) cmp cmp_val ()
+            elems' = case elems of
+                       [] -> error "Empty switch elements"
+                       [x] -> x
+                       (_:_) -> error "Only one switch element currently supported"
+        in (elems', equality)
+
+      pkSwitch :: String
+               -> String
+               -> TypeInfoMap
+               -> (String -> String)
+               -> (GenStructElem Type, Expr ())
+               -> (Expr (), Maybe (Expr ()))
+      pkSwitch name ext' m' _ (expr', cond) =
+        let acc' = \_ -> name
+            packed = structElemToPyPack ext' m' acc' expr'
+            -- Parsing these cases is annoying, there should be a better way to do this...
```

Comment:
Maybe it's time to make a real ADT for the result of structElemToPyPack so that we don't have to fuck around with all of these Nothing Right Lefts? (Also not necessary, because it's probably a lot of churn, but something to think about.)